File: ImperatorToCK3/Source/Mappers/TagTitleMapper/TagTitleMapping.cpp
#include "TagTitleMapping.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



mappers::TagTitleMapping::TagTitleMapping(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


std::optional<std::string> mappers::TagTitleMapping::tagRankMatch(const std::string& impTag, const std::string& rank) const {
	if (imperatorTag != impTag || (!ranks.empty() && !ranks.contains(rank)))
		return std::nullopt;

	return ck3Title;
}


void mappers::TagTitleMapping::registerKeys() {
	registerKeyword("ck3", [this](std::istream& theStream) {
		ck3Title = commonItems::getString(theStream);
	});
	registerKeyword("imp", [this](std::istream& theStream) {
		imperatorTag = commonItems::getString(theStream);
	});
	registerKeyword("rank", [this](std::istream& theStream) {
		ranks.emplace(commonItems::getString(theStream));
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: ImperatorToCK3/Mappers/TagTitle/TagTitleMapping.cs
using System.Collections.Generic;
using commonItems;

namespace ImperatorToCK3.Mappers.TagTitle {
	public class TagTitleMapping {
		public string? TagRankMatch(string imperatorTag, string rank) {
			if (this.imperatorTag != imperatorTag) {
				return null;
			}
			if (ranks.Count > 0 && !ranks.Contains(rank)) {
				return null;
			}
			return ck3Title;
		}

		private string ck3Title = string.Empty;
		private string imperatorTag = string.Empty;
		private readonly SortedSet<string> ranks = new();

		private static readonly Parser parser = new();
		private static TagTitleMapping mappingToReturn = new();
		static TagTitleMapping() {
			parser.RegisterKeyword("ck3", reader => {
				mappingToReturn.ck3Title = ParserHelpers.GetString(reader);
			});
			parser.RegisterKeyword("imp", reader => {
				mappingToReturn.imperatorTag = ParserHelpers.GetString(reader);
			});
			parser.RegisterKeyword("rank", reader => {
				mappingToReturn.ranks.Add(ParserHelpers.GetString(reader));
			});
			parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
		}
		public static TagTitleMapping Parse(BufferedReader reader) {
			mappingToReturn = new TagTitleMapping();
			parser.ParseStream(reader);
			return mappingToReturn;
		}
	}
}
